experiment/flake-detector/main.go

```diff
@@ -0,0 +1,220 @@
+	chance      float64
+	jobName     string
+}
+
+func (f jobResults) String() string {
+	return fmt.Sprintf("%d/%d\t(%.2f%%)\t%s", f.flakeCount, f.commitCount, f.chance, f.jobName)
+}
+
+type byChance []jobResults
+
+func (ft byChance) Len() int           { return len(ft) }
+func (ft byChance) Swap(i, j int)      { ft[i], ft[j] = ft[j], ft[i] }
+func (ft byChance) Less(i, j int) bool { return ft[i].chance > ft[j].chance }
+
+type options struct {
+	prowURL string
+	repo    string
+	path    string
+	runOnce bool
+}
+
+func flagOptions() options {
+	o := options{}
+	flag.StringVar(&o.prowURL, "prow-url", "https://deck-ci.svc.ci.openshift.org", "Prow frontend base URL. Required for reading job results")
+	flag.StringVar(&o.repo, "repo", "openshift/origin", "Repository to estimate test flakiness for")
+	flag.StringVar(&o.path, "config-path", "/etc/config/config.yaml", "Register prometheus metrics from the provided config")
+	flag.BoolVar(&o.runOnce, "once", false, "Run once and exit")
+	flag.Parse()
+	return o
+}
+
+func readHTTP(url string) ([]byte, error) {
+	var err error
+	retryDelay := time.Duration(2) * time.Second
+	for retryCount := 0; retryCount < 5; retryCount++ {
+		if retryCount > 0 {
+			time.Sleep(retryDelay)
+			retryDelay *= time.Duration(2)
+		}
+		resp, err := http.Get(url)
+		if resp != nil && resp.StatusCode >= 500 {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			continue
+		}
+		return body, nil
+	}
+	return nil, fmt.Errorf("ran out of retries reading from %q. The last error was: %v", url, err)
+}
+
+func sync(c *Config, o options) {
+	body, err := readHTTP(o.prowURL + "/data.js")
+	if err != nil {
+		log.Fatalf("error reading jobs from prow: %v", err)
+	}
+
+	var jobs []prowJob
+	if err := json.Unmarshal(body, &jobs); err != nil {
+		log.Fatalf("error unmarshaling prowjobs: %v", err)
+	}
+
+	jobMap := make(map[string]map[string][]string)
+	for _, job := range jobs {
+		if job.Type != "presubmit" {
+			continue
+		}
+		if job.Repo != o.repo {
+			continue
+		}
+		if job.State != "success" && job.State != "failure" {
+			continue
+		}
+		if _, ok := jobMap[job.Job]; !ok {
+			jobMap[job.Job] = make(map[string][]string)
+		}
+		if _, ok := jobMap[job.Job][job.PullSHA]; !ok {
+			jobMap[job.Job][job.PullSHA] = make([]string, 0)
+		}
+		jobMap[job.Job][job.PullSHA] = append(jobMap[job.Job][job.PullSHA], job.State)
+	}
+
+	jobCommits := make(map[string]int)
+	jobFlakes := make(map[string]int)
+	for job, commits := range jobMap {
+		jobCommits[job] = len(commits)
+		jobFlakes[job] = 0
+		for _, results := range commits {
+			hasFailure, hasSuccess := false, false
+			for _, state := range results {
+				if state == "success" {
+					hasSuccess = true
+				}
+				if state == "failure" {
+					hasFailure = true
+				}
+				if hasSuccess && hasFailure {
+					break
+				}
+			}
+			if hasSuccess && hasFailure {
+				jobFlakes[job]++
```

Comment:
Right now jUnit results are uploaded into BigQuery by Kettle to be analyzed for flakiness to identify the tests that cause the most amount of trouble. What niche do we expect to fill with this server?